pkg/template/event.go

```diff
@@ -18,177 +18,79 @@ package template
-		headerName := getHeaderFromVar(headerVar)
-		headerValue, err := getHeaderValue(header, headerName)
-		if err != nil {
-			return param, err
-		}
-		param.Value.StringVal = strings.Replace(param.Value.StringVal, headerVar, headerValue, -1)
-	}
-	return param, nil
-}
-
-// getHeaderValue returns a string representation of the headerName in the event
-// header. An error is returned if the headerName is not found in the header.
-func getHeaderValue(header map[string][]string, headerName string) (string, error) {
-	var headerValue string
-	if headerName == "" {
-		// $(header) has an empty headerName, so use all the headers in the headerValue
-		b, err := json.Marshal(&header)
-		if err != nil {
-			return "", xerrors.Errorf("Error marshalling header %s: %s", header, err)
-		}
-		headerValue = string(b)
-	} else {
-		value, ok := header[headerName]
-		if !ok {
-			return "", xerrors.Errorf("Error headerName %s not found in the event header %s", headerName, header)
-		}
-		headerValue = strings.Join(value, " ")
-	}
-	return strings.Replace(headerValue, `"`, `\"`, -1), nil
-}
-
 // ResolveParams takes a given trigger binding and produces the resulting
 // resource params.
-func ResolveParams(bindings []*triggersv1.TriggerBinding, body []byte, header map[string][]string, params []pipelinev1.ParamSpec) ([]pipelinev1.Param, error) {
+func ResolveParams(bindings []*triggersv1.TriggerBinding, body []byte, header http.Header, params []pipelinev1.ParamSpec) ([]pipelinev1.Param, error) {
 	out, err := MergeBindingParams(bindings)
 	if err != nil {
-		return nil, xerrors.Errorf("error merging trigger params: %v", err)
+		return nil, xerrors.Errorf("error merging trigger params: %q", err)
 	}
-	out, err = ApplyBodyToParams(body, out)
+	event, err := newEvent(body, header)
 	if err != nil {
-		return nil, xerrors.Errorf("error applying body to trigger params: %s", err)
+		return nil, xerrors.Errorf("failed to create Event: %q", err)
 	}
-	out, err = ApplyHeaderToParams(header, out)
+	out, err = applyEventValuesToParams(out, event)
 	if err != nil {
-		return nil, xerrors.Errorf("error applying header to trigger params: %s", err)
+		return nil, xerrors.Errorf("failed to ApplyEventValuesToParams: %q", err)
 	}
-
 	return MergeInDefaultParams(out, params), nil
 }
 
+// ResolveResources resolves a templated resource by replacing params with their values.
 func ResolveResources(template *triggersv1.TriggerTemplate, params []pipelinev1.Param) []json.RawMessage {
 	resources := make([]json.RawMessage, len(template.Spec.ResourceTemplates))
-	uid := UID()
 	for i := range template.Spec.ResourceTemplates {
 		resources[i] = ApplyParamsToResourceTemplate(params, template.Spec.ResourceTemplates[i].RawMessage)
-		resources[i] = ApplyUIDToResourceTemplate(resources[i], uid)
+		resources[i] = ApplyUIDToResourceTemplate(resources[i], UID())
 	}
 	return resources
 }
+
+// event represents a HTTP event that Triggers processes
+type event struct {
```

Comment:
might be confusing with `http.Request`. Also, I like event since we can later add the one-offs such as the `$(uid)` to this struct to unify how we do the "templating"

Reply:
> might be confusing with http.Request

it basically _is_ an http.Request tho since it has a header + a body :thinking: 

I see what you mean about the `uid` tho! you're probably right about `event` being better in the long run, sgtm :D